src/Carousel.js

```diff
@@ -63,233 +100,278 @@ const propTypes = {
 
 const defaultProps = {
   slide: true,
+  fade: false,
   interval: 5000,
+  keyboard: true,
   pauseOnHover: true,
   wrap: true,
   indicators: true,
   controls: true,
-  prevIcon: <Glyphicon glyph="chevron-left" />,
+  activeIndex: 0,
+
+  prevIcon: <span className="carousel-control-prev-icon" aria-hidden="true" />,
   prevLabel: 'Previous',
-  nextIcon: <Glyphicon glyph="chevron-right" />,
+
+  nextIcon: <span className="carousel-control-next-icon" aria-hidden="true" />,
   nextLabel: 'Next',
 };
 
 class Carousel extends React.Component {
   constructor(props, context) {
     super(props, context);
 
-    this.handleMouseOver = this.handleMouseOver.bind(this);
-    this.handleMouseOut = this.handleMouseOut.bind(this);
-    this.handlePrev = this.handlePrev.bind(this);
-    this.handleNext = this.handleNext.bind(this);
-    this.handleItemAnimateOutEnd = this.handleItemAnimateOutEnd.bind(this);
-
-    const { defaultActiveIndex } = props;
-
     this.state = {
-      activeIndex: defaultActiveIndex != null ? defaultActiveIndex : 0,
-      previousActiveIndex: null,
-      direction: null,
+      prevClasses: '',
+      currentClasses: 'active',
     };
-
     this.isUnmounted = false;
+
+    this.carousel = React.createRef();
   }
 
   componentDidMount() {
-    this.waitForNext();
+    this.cycle();
   }
 
-  componentWillReceiveProps(nextProps) {
-    const activeIndex = this.getActiveIndex();
-
-    if (
-      nextProps.activeIndex != null &&
-      nextProps.activeIndex !== activeIndex
-    ) {
-      clearTimeout(this.timeout);
+  static getDerivedStateFromProps(
+    nextProps,
+    { activeIndex: previousActiveIndex },
+  ) {
+    if (nextProps.activeIndex !== previousActiveIndex) {
+      const lastPossibleIndex =
+        ValidComponentChildren.count(nextProps.children) - 1;
+
+      const nextIndex = Math.max(
+        0,
+        Math.min(nextProps.activeIndex, lastPossibleIndex),
+      );
 
-      this.setState({
-        previousActiveIndex: activeIndex,
-        direction:
-          nextProps.direction != null
-            ? nextProps.direction
-            : this.getDirection(activeIndex, nextProps.activeIndex),
-      });
-    }
+      let direction;
+      if (
+        (nextIndex === 0 && previousActiveIndex >= lastPossibleIndex) ||
```

Comment:
when does this trigger? if previously there were 3 items, but now there's only 1? why should that trigger "next", v going from 3 -> 2 trigger "prev"?

Reply:
No, when the slides wrap you want to continue in the same direction you we're going. this isn't a perfect check but I think this component is dumb and don't want to spend lots of time on it :P

Reply:
hmm so that's if `previousActiveIndex == lastPossibleIndex`... except what if the user clicks on the indicator? or if you're cycling back from 0 to final